spring-data-cassandra/src/main/java/org/springframework/data/cassandra/config/CqlSessionFactoryBean.java

```diff
@@ -442,10 +469,22 @@ protected CqlSessionBuilder buildBuilder() {
 
 		Assert.hasText(this.contactPoints, "At least one server is required");
 
-		CqlSessionBuilder sessionBuilder = CqlSession.builder();
-
-		StringUtils.commaDelimitedListToSet(this.contactPoints).forEach(host ->
-			sessionBuilder.addContactPoint(InetSocketAddress.createUnresolved(host, this.port)));
+		CqlSessionBuilder sessionBuilder = setCqlSessionBuilderSupplier.get();
+
+		if (contactPointsWithPort == null) {
+			StringUtils.commaDelimitedListToSet(this.contactPoints)
+					.forEach(host -> sessionBuilder.addContactPoint(InetSocketAddress.createUnresolved(host, this.port)));
+		} else {
+			StringUtils.commaDelimitedListToSet(this.contactPointsWithPort).stream().map(HOST_PORT_SPLITTER::splitToList)
+					.forEach(hostAndPort -> {
+						if (hostAndPort.size() != 2) {
```

Comment:
If we decide to keep one `String` argument, then I think that we should do some validation. Right now I am doing a very simple check - if any `hostAndPort` is corrupted, the `IllegalArgumentException` with ALL `String contactPointsWithPort` is thrown. 
I can rework it to be more detailed, that is iterate over every pair of host and port, collect info about the corrupted ones, and report only those in the `IllegalArgumentException` message. Let me know wdyt about such detailed information - maybe the current validation is good enough?

Reply:
Argument validation should happen as early as possible, ideally in `setContactPoints` to not allow invalid values to get into the factory bean in the first place. `afterPropertiesSet()` is kind-of after the fact already and only good to check if required values weren't configured.

As discussed for backward compatibility, we should leniently accept host or host and port. We also need to check if the host contains an IPv6 as `:` is used in IPv6 addresses, so we'd need to check whether the address is using the resource identifier format with square brackets (`[2001:db8:85a3:8d3:1319:8a2e:370:7348]:443`) so that we correctly identify the port.